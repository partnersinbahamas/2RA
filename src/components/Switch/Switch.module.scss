@import '../utils/styles/styles.scss';
@import '../utils/styles/fonts.scss';

%base_input {
  appearance: none;
  cursor: pointer;
  border-radius: 0.75rem;
  margin: 0;

  border: 0.15em solid transparent;
  height: 1.55rem;
  width: 3rem;
  padding: 0.05rem 0.05rem;
  position: relative;
}

%base_line {
  content: '';
  position: absolute;
  top: 45%;
  right: 0.25rem;
  display: block;
  height: 1.5px;
  width: 0.85rem;
}

%base_check {
  content: '';
  box-sizing: border-box;
  display: block;
  width: 1.3rem;
  height: 1.2rem;
  border-radius: 50%;
  border: 2px solid transparent;
  transition: transform 0.15s;
  transform: translateX(2%);
}

.switch-wrapper {
  &[disabled] {
    cursor: not-allowed;

    * {
      cursor: inherit;
    }
  }
}

.switch-label {
  @extend %fontTuraRegular;
  font-size: 0.75rem;
  color: $cool_gray_600;
}

.switch-input {
  &-default[type='checkbox'] {
    @extend %base_input;

    border-color: $cool_gray_400;
    background-color: $cool_gray_200;

    &::before {
      @extend %base_check;
      background-color: $cool_gray_300;
      border-color: $cool_gray_400;
    }

    &:checked::before {
      transform: translateX(100%);
      background-color: $cool_gray_400;
    }

    &::after {
      @extend %base_line;
      background-color: $cool_gray_400;
    }

    &[disabled] {
      border-color: $cool_gray_400;
      background-color: $cool_gray_300;

      &::before {
        @extend %base_check;
        background-color: $cool_gray_400;
        border-color: $cool_gray_200;
      }

      &:checked::before {
        transform: translateX(100%);
        background-color: $cool_gray_400;
      }

      &::after {
        @extend %base_line;
        background-color: $cool_gray_200;
      }

      &:checked::after {
        @extend %base_line;
        left: 0.25rem;
        background-color: $cool_gray_200;
      }
    }
  }

  &-primary[type='checkbox'] {
    @extend %base_input;

    border-color: $blue_300;
    background-color: $blue_100;

    &::before {
      @extend %base_check;
      background-color: $blue_200;
      border-color: $blue_300;
    }

    &::after {
      @extend %base_line;
      background-color: $blue_300;
    }

    &:checked::before {
      background-color: $blue_300;
      transform: translateX(100%);
    }

    &[disabled] {
      border-color: $blue_400;
      background-color: $blue_300;

      &::before {
        @extend %base_check;
        background-color: $blue_400;
        border-color: $blue_200;
      }

      &:checked::before {
        transform: translateX(100%);
      }

      &::after {
        @extend %base_line;
        background-color: $blue_200;
      }

      &:checked::after {
        @extend %base_line;
        left: 0.25rem;
      }
    }
  }
}
