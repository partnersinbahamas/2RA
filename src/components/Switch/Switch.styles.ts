import styled from 'styled-components';

import { IHorizontal, IVertical } from '../utils/types/types';

export const Wrapper = styled.div<{
  labelPosition: IVertical | IHorizontal;
  disabled: boolean;
}>`
  width: min-content;
  display: flex;
  align-items: center;
  gap: 0.25rem;

  ${({ labelPosition }) => {
    switch (labelPosition) {
      case 'top':
        return `flex-direction: column`;
      case 'bottom':
        return `flex-direction: column-reverse`;
      case 'left':
        return `flex-direction: row`;
      case 'right':
        return `flex-direction: row-reverse`;
    }
  }}
`;

export const Label = styled.label<{ disabled: boolean }>``;
