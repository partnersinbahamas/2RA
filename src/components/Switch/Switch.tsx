import React, { useId, useState } from 'react';
import classNames from 'classnames';

import { useModuleExtention } from '../../lib';
import { useRA } from '../../providers/RAProvider';

import { IHorizontal, IVertical, TStile } from '../utils/types/types';
import defaultProps from '../utils/variables/defaultProps';

import styles from './Switch.module.scss';

type TProps = {
  label: string;
  required?: boolean;
  defaultToggle?: boolean;
  stile?: TStile;
  className?: any;
  labelPosition?: IHorizontal | IVertical;
  onClick?: () => void;
};

export const Switch: React.FC<TProps> = ({
  label,
  required = false,
  defaultToggle = false,
  stile = defaultProps.stile,
  className,
  labelPosition = 'top',
  onClick,
}) => {
  const id = `switch-${useId()}`;

  const [toggle, setToggle] = useState<boolean>(defaultToggle);
  const { stylesExtention, componentsStile } = useRA();
  const moduleExtention = useModuleExtention(
    stylesExtention as TStylesExtension,
  ).moduleExtentionState;

  const visibleStile = stile || componentsStile;

  const classes = {
    wrapper:
      className && (moduleExtention ? className['wrapper'] : `${className}`),
    label:
      className &&
      (moduleExtention ? className['label'] : `${className}-label`),
    input:
      className &&
      (moduleExtention ? className['input'] : `${className}-input`),
  };

  const stiles = {
    wrapper: styles[`switch-wrapper`],
    label: styles[`switch-label`],
    input: styles[`switch-input-${visibleStile}`],
  };

  const handleClick = () => {
    if (onClick) onClick();
    setToggle(current => !current);
  };

  return (
    <div
      className={classNames(
        classes.wrapper,
        stiles.wrapper,
        styles[`label-${labelPosition}`],
      )}
    >
      {label && (
        <label htmlFor={id} className={classNames(classes.label, stiles.label)}>
          {label}
          {required && '*'}
        </label>
      )}

      <input
        id={id}
        type="checkbox"
        required={required}
        className={classNames(classes.input, stiles.input)}
        onClick={handleClick}
        checked={toggle}
      />
    </div>
  );
};
