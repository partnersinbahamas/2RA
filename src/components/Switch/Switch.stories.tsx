import React from 'react';
import { Meta, StoryObj } from '@storybook/react';
import { Switch } from './Switch';
import { story } from '../utils/variables/story';
import { ArgTypes } from 'storybook/internal/types';
import { DocsPage, Subtitle } from '@storybook/blocks';

const meta = {
  component: Switch,
  title: 'Athomic/Switch',
  tags: ['autodocs'],
  args: {
    label: 'Switch',
  },
  argTypes: {
    onClick: story.onClick,
    label: story.title,
    defaultToggle: story.boolean,
    required: story.boolean,
    stile: story.stile,
    labelPosition: story.labelPosition,
  } as ArgTypes,
  parameters: {
    layout: 'centered',
    docs: {
      page: () => {
        return (
          <>
            <Subtitle>
              Please to style your custom Input component use the mockup bellow.
              below.
            </Subtitle>
            {/* <Markdown>Guide styles | .module.scss</Markdown>
            <Source language="css" code={INPUT_MODULE_SCSS_STYLES} />

            <Markdown>Guide styles | .scss</Markdown>
            <Source language="css" code={INPUT_SCSS_STYLES} /> */}
            <DocsPage />
          </>
        );
      },
    },
  },
} satisfies Meta<typeof Switch>;
export default meta;

type TStory = StoryObj<typeof meta>;
export const DefaultSwitch: TStory = {};
export const PrimarySwitch: TStory = { args: { stile: 'primary' } };
export const MuteSwich: TStory = { args: { stile: 'mute' } };
