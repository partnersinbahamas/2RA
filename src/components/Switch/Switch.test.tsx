import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Switch } from './Switch';

describe('Switch', () => {
  const user = userEvent.setup();

  const props = {
    label: 'Switch',
    onClick: jest.fn(),
  };

  it('renders correctly', () => {
    const { container } = render(<Switch />);

    const inputElement = screen.getByRole('checkbox');
    expect(inputElement).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('renders with label', () => {
    const { container } = render(<Switch {...props} />);

    const labelElement = screen.queryByText(props.label);
    expect(labelElement).toBeInTheDocument();

    const inputElement = screen.getByLabelText(props.label);
    expect(inputElement).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('renders with label position', () => {
    const { container } = render(<Switch {...props} labelPosition="bottom" />);

    const labelElement = screen.queryByText(props.label);
    expect(labelElement).toHaveAttribute('data-position', 'bottom');

    expect(container).toMatchSnapshot();
  });

  it('renders as required', () => {
    const labelAsRequired = `${props.label}*`;
    const { container } = render(<Switch {...props} required />);

    const labelElement = screen.queryByText(new RegExp(props.label));
    expect(labelElement?.textContent).toBe(labelAsRequired);

    const inputElement = screen.getByLabelText(labelAsRequired);
    expect(inputElement).toBeRequired();

    expect(container).toMatchSnapshot();
  });

  it('renders with stile', () => {
    const { container } = render(<Switch {...props} stile="primary" />);

    const parentElement = screen.queryByText(props.label)?.parentNode;
    expect(parentElement).toHaveAttribute('data-stile', 'primary');

    expect(container).toMatchSnapshot();
  });

  it('should be clicked', async () => {
    const { container } = render(<Switch {...props} />);

    const inputElement = screen.getByLabelText(props.label);
    await user.click(inputElement);
    expect(props.onClick).toHaveBeenCalled();

    expect(container).toMatchSnapshot();
  });

  it('should be disabled', async () => {
    const { container } = render(<Switch {...props} disabled />);

    const parentElement = screen.queryByText(props.label)?.parentElement;
    expect(parentElement).toHaveAttribute('disabled');

    const labelElement = screen.queryByText(props.label);
    expect(labelElement).toHaveAttribute('disabled');

    const inputElement = screen.getByLabelText(props.label);
    expect(inputElement).toBeDisabled();

    expect(container).toMatchSnapshot();
  });
});
