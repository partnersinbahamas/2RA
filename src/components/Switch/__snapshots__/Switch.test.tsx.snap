// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch renders as required 1`] = `
<div>
  <div
    class="switch-wrapper label-top"
    data-stile="default"
  >
    <label
      class="switch-label"
      data-position="top"
      for="switch-:r3:"
    >
      Switch
      *
    </label>
    <input
      class="sc-brSamD switch-input-default"
      id="switch-:r3:"
      required=""
      type="checkbox"
    />
  </div>
</div>
`;

exports[`Switch renders correctly 1`] = `
<div>
  <div
    class="switch-wrapper label-top"
    data-stile="default"
  >
    <input
      class="sc-brSamD switch-input-default"
      id="switch-:r0:"
      type="checkbox"
    />
  </div>
</div>
`;

exports[`Switch renders with label 1`] = `
<div>
  <div
    class="switch-wrapper label-top"
    data-stile="default"
  >
    <label
      class="switch-label"
      data-position="top"
      for="switch-:r1:"
    >
      Switch
    </label>
    <input
      class="sc-brSamD switch-input-default"
      id="switch-:r1:"
      type="checkbox"
    />
  </div>
</div>
`;

exports[`Switch renders with label position 1`] = `
<div>
  <div
    class="switch-wrapper label-bottom"
    data-stile="default"
  >
    <label
      class="switch-label"
      data-position="bottom"
      for="switch-:r2:"
    >
      Switch
    </label>
    <input
      class="sc-brSamD switch-input-default"
      id="switch-:r2:"
      type="checkbox"
    />
  </div>
</div>
`;

exports[`Switch renders with stile 1`] = `
<div>
  <div
    class="switch-wrapper label-top"
    data-stile="primary"
  >
    <label
      class="switch-label"
      data-position="top"
      for="switch-:r4:"
    >
      Switch
    </label>
    <input
      class="sc-brSamD switch-input-primary"
      id="switch-:r4:"
      type="checkbox"
    />
  </div>
</div>
`;

exports[`Switch should be clicked 1`] = `
<div>
  <div
    class="switch-wrapper label-top"
    data-stile="default"
  >
    <label
      class="switch-label"
      data-position="top"
      for="switch-:r5:"
    >
      Switch
    </label>
    <input
      class="sc-brSamD switch-input-default"
      id="switch-:r5:"
      type="checkbox"
    />
  </div>
</div>
`;

exports[`Switch should be disabled 1`] = `
<div>
  <div
    class="sc-csKJxZ jHCCer switch-wrapper label-top"
    data-stile="default"
    disabled=""
    labelposition="top"
  >
    <label
      class="sc-eTNRI switch-label"
      data-position="top"
      disabled=""
      for="switch-:r6:"
    >
      Switch
    </label>
    <input
      class="sc-brSamD switch-input-default"
      disabled=""
      id="switch-:r6:"
      type="checkbox"
    />
  </div>
</div>
`;

exports[`Switch should be rendered on error 1`] = `
<div>
  <div
    class="sc-csKJxZ jHCCer switch-wrapper label-top"
    data-stile="default"
    error="true"
    labelposition="top"
  >
    <label
      class="sc-eTNRI switch-label"
      data-position="top"
      error="true"
      for="switch-:r7:"
    >
      Switch
    </label>
    <input
      class="sc-brSamD switch-input-default"
      error="true"
      id="switch-:r7:"
      type="checkbox"
    />
  </div>
</div>
`;
